scalameta/dialects/shared/src/main/scala/scala/meta/Dialect.scala

```diff
@@ -0,0 +1,427 @@
+  // because sometimes dialects representing distinct Scala versions
+  // can be structurally equal to each other.
+  override def canEqual(that: Any): Boolean = this eq that.asInstanceOf[AnyRef]
+  override def equals(other: Any): Boolean = this eq other.asInstanceOf[AnyRef]
+  override def hashCode: Int = System.identityHashCode(this)
+  // Smart prettyprinting that knows about standard dialects.
+  override def toString = {
+    Dialect.standards.find(_._2 == this) match {
+      case Some((name, _)) => name
+      case None => s"Dialect(/* with custom parameters */)"
+    }
+  }
+
+  @deprecated("Use withX method instead", "4.3.11")
+  def copy(
+      allowAndTypes: Boolean = this.allowAndTypes,
+      allowAtForExtractorVarargs: Boolean = this.allowAtForExtractorVarargs,
+      allowCaseClassWithoutParameterList: Boolean = this.allowCaseClassWithoutParameterList,
+      allowColonForExtractorVarargs: Boolean = this.allowColonForExtractorVarargs,
+      allowEnums: Boolean = this.allowEnums,
+      allowImplicitByNameParameters: Boolean = this.allowImplicitByNameParameters,
+      allowImplicitFunctionTypes: Boolean = this.allowImplicitFunctionTypes,
+      allowInlineIdents: Boolean = this.allowInlineIdents,
+      allowInlineMods: Boolean = this.allowInlineMods,
+      allowLiteralTypes: Boolean = this.allowLiteralTypes,
+      allowMethodTypes: Boolean = this.allowMethodTypes,
+      allowMultilinePrograms: Boolean = this.allowMultilinePrograms,
+      allowOrTypes: Boolean = this.allowOrTypes,
+      allowPatUnquotes: Boolean = this.allowPatUnquotes,
+      allowSpliceUnderscores: Boolean = this.allowSpliceUnderscores,
+      allowTermUnquotes: Boolean = this.allowTermUnquotes,
+      allowToplevelTerms: Boolean = this.allowToplevelTerms,
+      allowTrailingCommas: Boolean = this.allowTrailingCommas,
+      allowTraitParameters: Boolean = this.allowTraitParameters,
+      allowTypeLambdas: Boolean = this.allowTypeLambdas,
+      allowViewBounds: Boolean = this.allowViewBounds,
+      allowWithTypes: Boolean = this.allowWithTypes,
+      allowXmlLiterals: Boolean = this.allowXmlLiterals,
+      toplevelSeparator: String = this.toplevelSeparator
+  ): Dialect = {
+    privateCopy(
+      allowAndTypes,
+      allowAtForExtractorVarargs,
+      allowCaseClassWithoutParameterList,
+      allowColonForExtractorVarargs,
+      allowEnums,
+      allowImplicitByNameParameters,
+      allowImplicitFunctionTypes,
+      allowInlineIdents,
+      allowInlineMods,
+      allowLiteralTypes,
+      allowMethodTypes,
+      allowMultilinePrograms,
+      allowOrTypes,
+      allowPatUnquotes,
+      allowSpliceUnderscores,
+      allowTermUnquotes,
+      allowToplevelTerms,
+      allowTrailingCommas,
+      allowTraitParameters,
+      allowTypeLambdas,
+      allowViewBounds,
+      allowWithTypes,
+      allowXmlLiterals,
+      toplevelSeparator
+    )
+  }
+}
+
+object Dialect extends InternalDialect {
+  def apply(
```

Comment:
should this also be deprecated?

Reply:
We may want to deprecate this method in the future but I think it's OK to keep it unchanged for now. I tried to introduce a default constructor `def apply(): Dialect` but couldn't find out a way to implement it without having to declare the default values in two different places, which I want to avoid. Most users adjust existing dialects like `Scala213`.